Turn off bundled subscription features

mgascam opened the pull request:
Fixes #WOOPMNT-5367

This pull request hides all merchant and customer-facing subscription management features in WooPayments, while keeping backend renewal processing working as usual.

In practice, this means that existing subscriptions handled by Stripe Billing will continue to renew automatically, but new subscriptions can no longer be created or managed from the WooCommerce interface.

#### Why this change?
WooPayments previously included a small, built-in version of subscription functionality. To reduce overlap and simplify the experience, we’re now retiring these built-in features and recommending the dedicated WooCommerce Subscriptions extension instead.

This transition helps us focus WooPayments on providing a seamless payments experience, while ensuring that merchants who rely on subscriptions have access to the full feature set and flexibility offered by WooCommerce Subscriptions.

#### Changes proposed in this Pull Request

- New disabler class
- Integration into subscriptions initialization
- Unit tests

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Test setup**
1. You should have existing active subscriptions to test:
  1. As a merchant, create a subscription product.
  2. As a shopper, complete a test purchase to create an active subscription.
2. Make sure you enable the bundle subscriptions feature flag in your test site. Go to WP Admin > WCPay Dev, check `Enable WCPay Subscriptions` and hit save. DO NOT enable the WooCommerce Subscriptions plugin.

**Hide WooCommerce > Subscriptions Page in wp-admin**
1. Log in as admin
2. Navigate to WooCommerce menu in wp-admin
3. Confirm "Subscriptions" submenu item is NOT visible under WooCommerce
4. Other WooCommerce menu items remain (Orders, Products, etc.)

<img width="339" height="261" alt="Screenshot 2025-10-20 at 18 52 22" src="https://github.com/user-attachments/assets/fa21abb9-e593-440c-aef3-9fd290fa5070" />|<img width="159" height="284" alt="Screenshot 2025-10-20 at 18 52 40" src="https://github.com/user-attachments/assets/95c7a502-971c-42aa-8294-1fca286869ec" />| 

**Direct URL Access Blocked (Custom Post Type)**
1. Attempt to access: `/wp-admin/edit.php?post_type=shop_subscription`
2. Note the redirect to `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments&wcpay_subscription_disabled=1`
3. No error messages displayed
4. User sees WooPayments settings page and this notice is displayed:
<img width="1104" height="109" alt="Screenshot 2025-10-28 at 11 46 22" src="https://github.com/user-attachments/assets/6bfc3a53-4ac4-4c3c-ab4a-31a33fd0b691" />


**Cannot Create Subscription Products**
1. Go to Products > Add New
2. Check the "Product type" dropdown
3. Confirm "Simple subscription" option is NOT present
4. Confirm "Variable subscription" option is NOT present
6. Confirm Regular product types still available (Simple, Variable, etc.)

<img width="477" height="304" alt="Screenshot 2025-10-20 at 19 00 02" src="https://github.com/user-attachments/assets/9582c483-0a67-46ff-8311-62a9361cd1d7" />

**Settings Tab Hidden**
1. Go to WooCommerce > Settings
2. Look through all available tabs
3. Confirm "Subscriptions" tab is NOT visible
4. Other tabs remain (General, Products, Shipping, etc.)

<img width="1083" height="113" alt="Screenshot 2025-10-20 at 19 10 19" src="https://github.com/user-attachments/assets/456f1ac9-82d1-4459-8e90-a9f23402aecb" />

**Direct Settings URL Blocked**
1. Attempt to access: `/wp-admin/admin.php?page=wc-settings&tab=subscriptions`
2. Note the redirect to: `/wp-admin/admin.php?page=wc-settings&tab=general`
3. Confirm General settings tab displayed and no error messages

**Cannot Edit Existing Subscriptions (List View)**
1. Create a test subscription (before checking out the branch)
2. Attempt to access: `/wp-admin/edit.php?post_type=shop_subscription`
3. Confirm you are redirected to WooPayments settings page and you cannot see the subscription list.

**Cannot Edit Existing Subscriptions (Single Edit)** 
1. Using a known subscription ID, attempt to access: `/wp-admin/post.php?post=[subscription_id]&action=edit`
2. Confirm you are redirected to WooPayments settings page

**My Account Menu Item Hidden**
1. Log in as a customer with an active subscription
2. Navigate to My Account page
3. Confirm "Subscriptions" link is NOT in the menu

<img width="1092" height="495" alt="Screenshot 2025-10-20 at 19 43 56" src="https://github.com/user-attachments/assets/ba10b674-7127-479b-a587-8ad1c9536b95" />

**Direct Subscriptions List URL Blocked**
1. While logged in as customer, attempt to access: `/my-account/subscriptions/`
2. Confirm you are redirected to to `/my-account/`

**Direct Subscription Detail URL Blocked**
1. Attempt to access: `/my-account/view-subscription/[subscription_id]/`
2. Confirm you are redirected to My Account dashboard

**Payment Method Change URL Blocked**
1. Attempt to access: `/my-account/subscription-payment-method/[subscription_id]/`
2. Confirm the redirect to My Account dashboard

**No Subscription Links in Orders**
1. Log in as customer
2. Go to My Account > Orders
3. Click on an order that created a subscription
4. View the order details page
5. Confirm "Related Subscriptions" section is NOT displayed
7. Order details should show normally

<img width="750" height="699" alt="Screenshot 2025-10-20 at 19 57 26" src="https://github.com/user-attachments/assets/3b17baab-8e92-4de0-a063-9de2d2e704ae" />

**Admin Order View (Customer Perspective)**
1. As admin, view an order in wp-admin that has a related subscription
2. Check for "Related Subscriptions" meta box
3. Confirm "Related Subscriptions" section is removed

**Automated Renewal Processing (Webhook-Based)**
Note: I could not make this work on my local, only with JN.
1. Identify an active subscription. Note the number of renewal orders in WooCommerce
2. In Stripe Dashboard (test mode), locate the subscription and click it
3. Locate a "Start Simulation" button on the top right corner of the sceen and click it.
4. Use the modal that appears to advance time to the scheduled renewal (Sometimes I had to add one or two more days)
5. Confirm new renewal order created in the test site
6. Confirm renewal order marked as "Processing"
8. Customer receives order confirmation email

<img width="281" height="181" alt="Screenshot 2025-10-23 at 18 22 47" src="https://github.com/user-attachments/assets/9dd09f4c-71d2-4591-81c5-e3b712d5763d" />



**Failed Renewal Handling** 
1.  Create a test subscription with valid payment method (using a different branch)
2. Change the payment method of the subscription to an invalid payment method (`4000 0000 0000 0341`works for this purpose) or manually fail an invoice in Stripe Dashboard
3. Trigger renewal in Stripe using the simulation tool described in the previous step
4. Confirm renewal order created (in "Failed" status)
5. Subscription status updated (On hold)
6. Customer receives failed payment notification
7. Order notes reflect failure reason

<img width="295" height="500" alt="Screenshot 2025-10-23 at 18 21 55" src="https://github.com/user-attachments/assets/821702d2-f15f-4a1b-b486-44315abcee1e" />

**Prevent creation of new bundled subscription orders (Shopper)**
1. As a shopper, navigate to the product list in the Shop page
2. Locate a subscription product in the list
3. Confirm the 'Add to Cart' button is not available, only the Read More button.
4. Navigate to the product details page of a subscription product
5. Confirm the product cannot be added to the cart.

<img width="411" height="481" alt="Screenshot 2025-10-27 at 10 45 22" src="https://github.com/user-attachments/assets/c08d83d1-604b-44f7-b114-4e01fb21ee32" />

**Prevent creation of new bundled subscription orders (Merchant)**
1. As a merchant, navigate to WooCommerce > Orders and click the `Add Order`
2. Click `Add Item` and then `Add Product`
3. Confirm the bundled subscription product does not show up on the `Add Products` modal search.

**Prevent enabling bundled subscriptions feature (Merchant)**
1. Navigate to Payments > Settings > Advanced settings
2. Locate the "WooPayments Subscriptions" toggle (should be ON/enabled)
3. Toggle the switch OFF and save settings
4. Confirm settings save successfully.
5. Try to toggle the switch back ON
6. Confirm the checkbox appears disabled out and cannot be turned back ON
7. Help text shows: "This feature has been deprecated. To manage subscriptions, please install WooCommerce Subscriptions."

<img width="986" height="182" alt="Screenshot 2025-10-28 at 11 53 23" src="https://github.com/user-attachments/assets/0aa07cad-4776-4e5c-9495-b22db365b334" />

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-10.2.0#turn-off-bundled-subscription-features
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Add WC_Payments_Subscriptions_Disabler class and related unit tests

mgascam committed:
Add settings tab removal and redirection for subscriptions in WC_Payments_Subscriptions_Disabler

mgascam committed:
Add functionality to remove related subscriptions section from order details and corresponding unit test

mgascam committed:
Refactor hook priorities in WC_Payments_Subscriptions_Disabler for consistency

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

mgascam committed:
Add doc comments

mgascam committed:
Add unit test

mgascam committed:
Prevent editin existing subscription

mgascam committed:
Fix test

mgascam committed:
Add changelog

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

rtio reviewed (commented):
When I was testing `Cannot Edit Existing Subscriptions (Single Edit)` I could enter the edit page but the simple subscription had become a simple product. Is that expected?

rtio reviewed (changes requested):
I also failed following those instructions `Automated Renewal Processing (Webhook-Based)` the subscription appeared as canceled on the Stripe dashboard. And `Failed Renewal Handling` same reason.

Another thing is that, after installing the actual subscriptions plugin all the functionality remained hidden.

rtio commented on includes/subscriptions/class-wc-payments-subscriptions-disabler.php:
```diff
@@ -0,0 +1,348 @@
+	 */
+	protected function redirect( $target ) {
+		wp_safe_redirect( $target );
+		exit;
```
do you actually have to exit here?

mgascam replied:
I think so, I wasn't completely sure so I checked and it's mentioned in the [docs](https://developer.wordpress.org/reference/functions/wp_safe_redirect/)

mgascam committed:
Disable bundled subscriptions UI only when WooCommerce Subscriptions plugin is inactive

mgascam committed:
Block editing of subscription products and ensure regular products are not affected

mgascam committed:
Add admin notice for disabled subscription features and update redirect URL

mgascam commented:
Thanks for the review and testing @rtio. I addressed some of your comments:

> When I was testing Cannot Edit Existing Subscriptions (Single Edit) I could enter the edit page but the simple subscription had become a simple product. Is that expected?

This was indeed confusing, so I changed it to redirect to the settings screen and show a notice like this:

<img width="1844" height="88" alt="Screenshot 2025-10-23 at 13 33 21" src="https://github.com/user-attachments/assets/12d3f9de-f732-4418-bed2-dc329e2b27c8" />

> Another thing is that, after installing the actual subscriptions plugin all the functionality remained hidden.

Great catch, I was planning to check for the WooCommerce Subscriptions but it slipped my mind 🤦 . Now it checks before disabling.

> I also failed following those instructions Automated Renewal Processing (Webhook-Based) the subscription appeared as canceled on the Stripe dashboard. And Failed Renewal Handling same reason.

I used a JN to test this part and it seemed to work for me, not sure if webhooks maybe? Anyway I'm going to test again and let you know when it's ready.

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

mgascam commented:
@rtio I tested again the renewal flows and they worked for me, please check the linear issue for more details. It's ready for a new review round when you have a moment. Thanks

rtio reviewed (commented):
Thanks for addressing the comments. I have a couple of other:

- Just a comment, there are a few E2E tests failing;
- When testing this `Cannot Edit Existing Subscriptions (Single Edit)` instead of being redirected to `WooCommerce overview` I got redirected to `wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments&wcpay_subscription_disabled=1`. But I confirm I saw the following message: `Subscription management is currently unavailable. To create and manage subscriptions, please install [WooCommerce Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/).`
- When testing `Cannot Edit Existing Subscriptions (List View)` I also got redirected to `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments&wcpay_subscription_disabled=1`

leonardola reviewed (changes requested):
I followed these steps but the pay for order flow showed up instead of me getting forwarded to My Account Dashboard.

> Payment Method Change URL Blocked
> - Attempt to access: /my-account/subscription-payment-method/[subscription_id]/
> - Confirm the redirect to My Account dashboard

I followed these steps but I'm still seeing the `Related Orders` box which contains a subscription link. I'm not sure if that's the section that should be hidden though.

> Admin Order View (Customer Perspective)
> - As admin, view an order in wp-admin that has a related subscription
> - Check for "Related Subscriptions" meta box
> - Confirm "Related Subscriptions" section is removed

When I visit the /shop page I see the subscription product listed and it still has the info about renewal periods:
<img width="460" height="695" alt="Image" src="https://github.com/user-attachments/assets/ee814f1a-941b-41af-a6ee-086a840ca0a4" />

After adding the product to the cart and visiting the checkout page I also see renewal info: 
<img width="496" height="643" alt="Image" src="https://github.com/user-attachments/assets/dea102a0-3bde-4be1-a2bf-b01ec3d055dc" />

I was able to place the order and a subscription got created.
Steps followed:
- Checkout to the current branch
- Visit the /shop page
- See the subscription is still showing up with it's renewal period
- Add the product to your cart
- Visit the cart page
- Notice the subscription renewal info
- Go to the checkout page
- Notice the subscription renewal info
- Place the order
- A new order and a new subscription are created

mgascam committed:
Add inbox deprecation notice

mgascam committed:
Update notice copy

mgascam committed:
Add functionality to prevent purchasing of subscription products

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

mgascam committed:
Filtering subscription products from admin search and blocking their addition to orders

mgascam committed:
Prevent direct access to order-pay endpoints for subscription IDs.

mgascam committed:
Remove "Related Orders" meta box from order edit screens

mgascam commented:
Thanks for the comments and the testing folks @leonardola @rtio. Now it is ready for a new review round.

Please find more details in [Linear](https://linear.app/a8c/issue/WOOPMNT-5367/turn-off-bundled-subscription-features#comment-4219f19c)

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

leonardola commented on tests/unit/notes/test-class-wc-payments-notes-stripe-billing-deprecation.php:
```diff
@@ -31,9 +31,12 @@ public function test_get_note() {
 			$title   = 'WooPayments subscriptions update';
 			$content = 'WooPayments no longer supports billing for existing customer subscriptions. All subscriptions data is read-only. Please install WooCommerce Subscriptions to continue managing your subscriptions.';
-		} else {
+		} elseif ( version_compare( $wcpay_version, '10.2.0', '<' ) ) {
```
This will never evaluate to true right? We're launching the code on 10.2.0 so it will never be < 10.2.0

mgascam replied:
Done here [28bd02c](https://github.com/Automattic/woocommerce-payments/pull/11089/commits/28bd02c1ff9d1fdd6eea2dfecfee2acebda8152c)

leonardola commented on includes/notes/class-wc-payments-notes-stripe-billing-deprecation.php:
```diff
@@ -63,9 +63,12 @@ public static function get_note() {
 			$note->set_title( __( 'WooPayments subscriptions update', 'woocommerce-payments' ) );
 			$note->set_content( __( 'WooPayments no longer supports billing for existing customer subscriptions. All subscriptions data is read-only. Please install WooCommerce Subscriptions to continue managing your subscriptions.', 'woocommerce-payments' ) );
-		} else {
+		} elseif ( version_compare( $wcpay_version, '10.2.0', '<' ) ) {
```
This will never evaluate to true right? We're launching the code on 10.2.0 so it will never be < 10.2.0

mgascam replied:
Good catch! You are absolutely right. I also think it makes sense to remove the other version checks. Since we're doing a complete removal in 10.2.0, all those checks are now dead code that would never evaluate to true. 

Done here [28bd02c](https://github.com/Automattic/woocommerce-payments/pull/11089/commits/28bd02c1ff9d1fdd6eea2dfecfee2acebda8152c)

leonardola reviewed (changes requested):
In /wp-admin > Payments > Settings > Advanced settings we need prevent people from enabling it from now on

<img width="986" height="319" alt="Image" src="https://github.com/user-attachments/assets/1615d72a-49a9-4355-a0cf-c1e39044d5e4" />

rtio reviewed (commented):
Code LGTM, besides Leonardo's comments everything seems fine. We should fix this https://github.com/Automattic/woocommerce-payments/pull/11089#pullrequestreview-3385020117

mgascam committed:
Merge branch 'develop' into update/woopmnt-5367-turn-off-bundled-subscription-features

mgascam committed:
Refactor Stripe Billing deprecation note to simplify message handling and remove version checks

mgascam committed:
Deprecate WooCommerce Payments subscriptions feature and prevent re-enabling in settings

mgascam committed:
Refactor WCPay subscriptions tests to use 'default_option_' filter for better control over feature flag

mgascam commented:
Thanks again for the review @leonardola @rtio, latest comments are addressed now. Waiting for approvals or any other feedback.

mgascam committed:
Filter the message when subscriptions products are
removed from the cart.

rtio reviewed (commented):
Question, should I still see subscription products?

<img width="2438" height="973" alt="Image" src="https://github.com/user-attachments/assets/83ed3079-b5e0-4b5f-9a5b-36720b8e6847" />

rtio reviewed (approved):
Besides my question, I'm approving this PR. Let's wait @leonardola's review before merging.

mgascam commented:
> Question, should I still see subscription products?

@rtio Yes, subscription products can still be listed in the shop, but checkout is disabled. Please see [this Linear comment](https://linear.app/a8c/issue/WOOPMNT-5367/turn-off-bundled-subscription-features#comment-1e0f636b)
 for context.

leonardola reviewed (approved):
Changes look good to me. I found only one minor when trying to disable WCPay subscriptions:

https://github.com/user-attachments/assets/538f022a-c259-4e13-8ccd-8f66720c96b7

mgascam commented:
> Changes look good to me. I found only one minor when trying to disable WCPay subscriptions:
> 
>  Screen.Recording.2025-10-28.at.17.39.31.mov

@leonardola I couldn’t reproduce the issue, either locally or on the JN site 🤔. The only case I found where the “Save changes” button stays disabled is when another required field (like the Support email) is empty. In that situation, I’d say the behavior is expected.
I’ll go ahead and merge since this looks like a minor issue, but please let me know if you have any additional details. The Redux state or any console errors would be super helpful, since it seems to be a JS-side issue. Thanks!

leonardola commented:
> button stays disabled is when another required field (like the Support email) is empty

You're absolutely right I missed those fields. Now I'm able to click the save button but when I reload the page it gets checked again. 

https://github.com/user-attachments/assets/3c28130d-0d17-40db-b6b2-29a539ed6451

mgascam commented:
> Now I'm able to click the save button but when I reload the page it gets checked again.

🤦 Thanks @leonardola, [this should fix it](https://github.com/Automattic/woocommerce-payments/pull/11113)

htdat commented:
Sounds like this issue also should be closed? https://github.com/Automattic/woocommerce-payments/issues/10403 

_Edit_: I've closed it. Feel free to revert the action if I am doing something incorrect.
